### PR TITLE
TELCODOCS-416: CNF-3205 Telco friendly workload specific PerformanceProfile API WIP

### PR DESCRIPTION
**Pull request by tmulquee**
Version(s):
4.11 and onwards.
Issue:
https://issues.redhat.com/browse/TELCODOCS-416

Issue:
https://issues.redhat.com/browse/TELCODOCS-416

First mention in main body of docs. Final paragraph of "Understanding law latency" concept topic.

http://file.emea.redhat.com/tmulquee/TELCODOCS-416/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html
Topic on Configuring workload hints: planning issues:

http://file.emea.redhat.com/tmulquee/TELCODOCS-416/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#cnf-configuring-workload-hints_cnf-master
Sample output from cat yaml now includes workflow hints (step 6):

http://file.emea.redhat.com/tmulquee/TELCODOCS-416/scalability_and_performance/cnf-create-performance-profiles.html#running-the-performance-profile-profile-cluster-using-podman_cnf-create-performance-profiles

Topic on arguments for Performance Profile (power-consumption-mode row in table):

http://file.emea.redhat.com/tmulquee/TELCODOCS-416/scalability_and_performance/cnf-create-performance-profiles.html#performance-profile-creator-arguments_cnf-create-performance-profiles

### CODE REVIEW

**Review by marioferh — COMMENTED** [review]
cool, lgtm. If  @MarSik has any further suggestion...

File: modules/cnf-configuring-workload-hints.adoc [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Workload hints simplify the fine-tuning of performance to industry sector settings. Instead of a “one size fits all” approach, workload hints can cater to usage patterns such as placing priority on:
+
+* Low latency
+* Real-time capability
+* Efficient use of power
+* High throughput
+
+In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload class and better able to meet the demands of a workload class. The cluster admin can now specify into which use case a workload falls. The Node Tuning Operator uses hints fine tune the performance settings.
+
+The environment in which an application is operating influences the choice of hints. In settings such as a core network, only minimal default power tuning is supported using CPU partitioning. In  distributed or Edge settings, latency is a higher priority, but some measures are still taken to optimize power consumption. For far edge settings, at a plane where where software-based radio is transmitted, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.
```
Comment (yanirq) [inline modules/cnf-configuring-workload-hints.adoc:18]: > only minimal default power tuning is supported using CPU partitioning
@MarSik  can we state that this is the most common case ?

Comment (yanirq) [inline modules/cnf-configuring-workload-hints.adoc:18]: > where where
@tmulquee  remove one `where`

File: modules/cnf-performance-profile-creator-arguments.adoc [hunk 2]
```diff
@@ -52,9 +52,9 @@ a|The power consumption mode.
 
 Possible values:
 
-* `default`
-* `low-latency`
-* `ultra-low-latency`
+* `default`: CPU partitioning with enabled power management and basic low-latency.
```
Comment (yanirq) [inline modules/cnf-performance-profile-creator-arguments.adoc:55]: I wonder if we need to mention that default means setting no value or is this a given.

File: modules/ipi-preparing-reinstall-cluster-bare-metal.adoc [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+// //installing/installing_bare_metal_ipi/installing_bare_metal_ipi/ipi-install-installation-workflow.adoc
+
+[id="ipi-preparing-reinstall-cluster-bare-metal_{context}"]
+
+= Preparing to reinstall a cluster on bare-metal
```
Comment (yanirq) [inline modules/ipi-preparing-reinstall-cluster-bare-metal.adoc:6]: is this file related to the context of this PR ?

File: modules/cnf-configuring-workload-hints.adoc [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+[id="cnf-configuring-workload-hints_{context}"]
+= Configuring workload hints
+
+Workload hints simplify the fine-tuning of performance to industry sector settings. Instead of a “one size fits all” approach, workload hints can cater to usage patterns such as placing priority on:
+
+* Low latency
+* Real-time capability
+* Efficient use of power
+
+In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload class and better able to meet the demands of a workload class. The cluster admin can now specify into which use case a workload falls. The Node Tuning Operator uses hints fine tune the performance settings.
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:15]:
```suggestion
In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload expectations and better able to meet the demands of the workload. The cluster admin can now specify into which use case his workload falls. The Node Tuning Operator uses hints to fine tune the performance settings for it.
```

File: modules/cnf-configuring-workload-hints.adoc (hunk 1 above)
Comment (MarSik) [reply modules/cnf-configuring-workload-hints.adoc:17]: The paragraph needs rewrite, it is jumping from one undefined term to another, mixing concepts etc..

I propose something like this (I edited out all the Telco specific terms that were not defined anywhere):

```suggestion
The environment in which an application is operating influences its behavior. For a typical datacenter with no strict latency requirements, only minimal default tuning is needed that enables CPU partitioning for the few and far between high performance workload pods.
Then there are data centers and workloads where latency is a higher priority, but some measures are still taken to optimize power consumption.
The most complicated cases are clusters close to latency sensitive pieces of equipment like manufacturing machinery, software defined radios etc. This last class of deployment is also known as Far edge. For far edge deployments, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    | Hint | Application | Environment | Description
+
+    | Default
+    a|[source,terminal]
+----
+workloadHints:
+highPowerConsumption: false
+realtime: false
+----
+    | Core network operations
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:30]:
```suggestion
    | High throughput cluster without latency requirements
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload class and better able to meet the demands of a workload class. The cluster admin can now specify into which use case a workload falls. The Node Tuning Operator uses hints fine tune the performance settings.
+
+The environment in which an application is operating influences the choice of hints. In settings such as a core network, only minimal default power tuning is supported using CPU partitioning. In  distributed or Edge settings, latency is a higher priority, but some measures are still taken to optimize power consumption. For far edge settings, at a plane where where software-based radio is transmitted, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.
+
+[cols="1,1,1,1",options="header"]
+|===
+    | Hint | Application | Environment | Description
+
+    | Default
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:23]: This column needs a header pointing to the Performance profile creator power mode

File: modules/cnf-configuring-workload-hints.adoc [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+
+    | Low-latency
+    a|[source,terminal]
+----
+workloadHints:
+highPowerConsumption: false
+realtime: true
+----
+    | Edge applications
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:42]: Workloads with modest latency requirements, regional datacenters. I would stay away from simple "Edge", because it is ambiguous for non-Telco folks.

File: modules/cnf-configuring-workload-hints.adoc [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+
+    | Ultra-low-latency
+    a|[source,terminal]
+----
+workloadHints:
+highPowerConsumption: true
+realtime: true
+----
+    | Far edge settings
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:53]:
```suggestion
    | Far edge clusters, latency critical workloads
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    | No power management, minimal latency, and maximum determinism.
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:54]:
```suggestion
    | Optimized for absolute minimal latency and maximum determinism at the cost of increased power consumption.
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+|===
+
+.Procedure
+
+. Create a profile appropriate for the environment's hardware and topology. In this example, we tune for ultra-low-latency.
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:60]:
```suggestion
. Create a PerformanceProfile appropriate for the environment's hardware and topology as described in chapters above. Then tweak the profile to match the expected workload. In this example, we tune for the lowest possible latency.
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+. Add the `highPowerConsumption` and `realtime` parameters for ultra-low-latency workload hints:
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:62]:
```suggestion
. Add the `highPowerConsumption` and `realtime` workload hints and set them both to true according to the table above.
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    kind: PerformanceProfile
+    metadata:
+      name: workload-hints
+    spec:
+      ...
+      workloadHints:
+        highPowerConsumption: true <1>
+        realtime: true <2>
+----
+<1> If `highPowerConsumption` is true, set very low latency at the cost of increased power consumption.
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:76]:
```suggestion
<1> If `highPowerConsumption` is true, the node is tuned for very low latency at the cost of increased power consumption.
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+<2> Disables certain debugging and monitoring features, but leaves power management on.
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:77]:
```suggestion
<2> Disables certain debugging and monitoring features that can affect system latency.
```

File: modules/cnf-configuring-workload-hints.adoc (hunk 1 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:17]: Thanks for the re-write of my first draft. I made some changes as required by our Style Guide.

Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:17]: Done.

File: modules/ipi-preparing-reinstall-cluster-bare-metal.adoc (hunk 3 above)
Comment (tmulquee) [reply modules/ipi-preparing-reinstall-cluster-bare-metal.adoc]: I've removed the file. It was included by accident.

File: modules/cnf-configuring-workload-hints.adoc (hunk 4 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:15]: Done with minor style guide changes - such as "his" -> "that".

File: modules/cnf-configuring-workload-hints.adoc (hunk 6 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:23]: I suggest preceding the table with a Note: "The following workload hints can be configured manually or created using the Performance Profile Creator. See "Creating a performance profile" for further information."

File: modules/cnf-configuring-workload-hints.adoc (hunk 8 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:53]: Done.

File: modules/cnf-configuring-workload-hints.adoc (hunk 9 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:54]: Done.

File: modules/cnf-configuring-workload-hints.adoc (hunk 10 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:60]: Done.

File: modules/cnf-configuring-workload-hints.adoc (hunk 11 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:62]: Done.

File: modules/cnf-configuring-workload-hints.adoc (hunk 13 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:77]: Done.

File: modules/cnf-configuring-workload-hints.adoc [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    | Far edge clusters, latency critical workloads
+    | Optimized for absolute minimal latency and maximum determinism at the cost of increased power consumption.
+
+|===
+
+.Procedure
+
+. Create a `PerformanceProfile` appropriate for the environment's hardware and topology as described in previous sections. Then adjust the profile to match the expected workload. In this example, we tune for the lowest possible latency.
+
+. Add the `highPowerConsumption` and `realtime` workload hints and set them both to true as indicated by the above table.
```
Comment (marioferh) [inline modules/cnf-configuring-workload-hints.adoc:69]: in the .md is realTime:
`| realTime | RealTime defines if the node should be configured for the real time workload. | *bool | false |`

but in the yaml is realtime. Not sure why highPowerConsumption is camelCase and realtime is not.

**Review by marioferh — APPROVED** [review]
/lgtm

File: modules/cnf-configuring-workload-hints.adoc (hunk 14 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:69]: If realtime in the yaml I will leave as is.

File: modules/cnf-configuring-workload-hints.adoc [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+The most complicated cases are clusters close to latency-sensitive equipment such as manufacturing machinery and software-defined radios. This last class of deployment is often referred to as Far edge. For Far edge deployments, ultra-low latency is the ultimate priority, and is achieved at the expense of power management.
+
+[NOTE]
+====
+The following workload hints can be configured manually or created using the Performance Profile Creator. See "Creating a performance profile" for further information.
+====
+
+[cols="1,1,1,1",options="header"]
+|===
+    | `PerformanceProfile` setting| Hint | Environment | Description
```
Comment (MarSik) [inline modules/cnf-configuring-workload-hints.adoc:28]:
```suggestion
    | Performance Profile creator setting | Hint | Environment | Description
```

File: modules/cnf-configuring-workload-hints.adoc [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+[id="cnf-configuring-workload-hints_{context}"]
+= Configuring workload hints
+
+Workload hints simplify the fine-tuning of performance to industry sector settings. Instead of a “one size fits all” approach, workload hints can cater to usage patterns such as placing priority on:
+
+* Low latency
+* Real-time capability
+* Efficient use of power
+
+In an ideal world, all of those would be prioritized: in real life, some come at the expense of others. The Node Tuning Operator is now aware of the workload expectations and better able to meet the demands of the workload. The cluster admin can now specify into which use case that workload falls. The Node Tuning Operator uses hints to fine tune the performance settings for the workload.
```
[the remark this answers is not in the document]
Comment (MarSik) [reply modules/cnf-configuring-workload-hints.adoc:15]: Or did we say The NTO uses PerformanceProfile? Makes little difference to me though.

File: modules/cnf-configuring-workload-hints.adoc (hunk 15 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc]: Done.

File: modules/cnf-configuring-workload-hints.adoc (hunk 16 above)
[the remark this answers is not in the document]
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc]: Went with: "The Node Tuning Operator uses `PerformanceProfile` to fine tune the performance settings for the workload."

**Comment by tmulquee** [comment]
I've implemented the suggested changes, which removed the LGTM label. Please re-add the LGTM when verified.

**Comment by tmulquee** [comment]
I've implemented the suggested changes, which removed the LGTM label.
Please re-add the LGTM when verified.

File: modules/cnf-configuring-workload-hints.adoc [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    metadata:
+      name: workload-hints
+    spec:
+      ...
+      workloadHints:
+        highPowerConsumption: true <1>
+        realtime: true <2>
+----
+<1> If `highPowerConsumption` is true, the node is tuned for very low latency at the cost of increased power consumption.
+<2> Disables some debugging and monitoring features that can affect system latency.
```
Comment (johnwilkins) [inline modules/cnf-configuring-workload-hints.adoc:84]: I think this module needs to be restructured or broken up into multiple modules. The module begins as a concept. Then, it provides a table of reference materials. Finally, it has a procedure. 

I would also add an ==Additional resources section after the procedure with a link to "Creating a performance profile," because the first step of the procedure says "as described in previous sections", and it isn't in a previous section. It's actually in the subsequent assembly.

Let me know if you need help with this.

**Review by johnwilkins — CHANGES_REQUESTED** [review]
As a general comment, from the introduction of the bulleted priorities in the workload hints module, the bullets differ from the actual settings, which I found a bit confusing. I was expecting profiles with names like those bullets.

**Comment by tmulquee** [comment]
Hi John
Thanks for the feedback, it's much appreciated, especially for brand new
material that's still taking shape.
I'll set about implementing the suggestions and I'll call on you if I run
into any difficulty.
Best regards
/tonym

File: modules/cnf-configuring-workload-hints.adoc (hunk 17 above)
Comment (tmulquee) [reply modules/cnf-configuring-workload-hints.adoc:84]: As mentioned above, I relocated the conceptual material to the top level conceptual topic on low latency ("Understanding low latency"). I gave the table a topic of its own - it needs to precede the procedure, as you have to use the table to know the right workload hint settings to go for.

File: modules/cnf-understanding-low-latency.adoc [hunk 18]
```diff
@@ -42,8 +42,10 @@ set values, installing a kernel, and reconfiguring the machine. But this method
...
 that, when done manually, is complex and could be prone to mistakes.
 
-{product-title} provides a Performance Addon Operator to implement automatic
+{product-title} now uses the Node Tuning Operator to implement automatic
 tuning to achieve low latency performance for OpenShift applications.
 The cluster administrator uses this performance profile configuration that makes
 it easier to make these changes in a more reliable way. The administrator can
 specify whether to update the kernel to kernel-rt, reserve CPUs for cluster and operating system housekeeping duties, including pod infra containers, and isolate CPUs for application containers to run the workloads.
+
+{product-title} also supports a hinting mechanism for the Node Tuning Operator that can tune `PerformanceProfile` to meet the demands of different industry environments. In this release, workload hints are available for `highPowerConsumption` (very low latency at the cost of increased power consumption) and `realtime` (priority given to optimum latency). A combination of true/false settings for these hints can be used to deal with application-specific workload profiles and requirements.
```
[the remark this answers is not in the document]
Comment (tmulquee) [reply modules/cnf-understanding-low-latency.adoc:51]: I use the article in all instances of PerformanceProfile, and applied backticks to true|false settings.
Removed RN-type usage.
